internal/services/oidc.go

```diff
@@ -0,0 +1,250 @@
+	}
+}
+
+// GetReadyzFunc make sure if oidc provider is ready.
+func (osvc *OIDCService) GetReadyzFunc() func() bool {
+	return osvc.client.ReadyzFunc
+}
+
+// GetOIDCProviderURL gets OIDC provider URL from the OAuth2 configuration.
+func (osvc *OIDCService) GetOIDCProviderURL(w http.ResponseWriter, r *http.Request) string {
+	// Sets a temporary session with state and nonce to validate
+	// when there is callback from the OIDC Provider.
+	session := osvc.sessionStorage.CreateSession()
+
+	return osvc.client.OAuth2Config.AuthCodeURL(
+		session.State,
+		oidc.Nonce(session.Nonce),
+	)
+}
+
+// GetUserInfo gets User Info from the OAuth2 Token.
+func (osvc *OIDCService) GetUserInfo(ctx context.Context, oauth2Token *oauth2.Token) (*oidc.UserInfo, error) {
+	userInfo, err := osvc.client.Provider.UserInfo(ctx, oauth2.StaticTokenSource(oauth2Token))
+	if err != nil {
+		return nil, err
+	}
+
+	return userInfo, nil
+}
+
+// GetValueFromToken gets the nonce from the ID Token.
+func (osvc *OIDCService) GetValueFromToken(value string, t *models.Tokens) (interface{}, error) {
+	var m map[string]interface{}
+	if err := json.Unmarshal(*t.IDTokenClaims, &m); err != nil {
+		return "", err
+	}
+
+	if key, ok := m[value]; ok {
+		return key, nil
+	}
+	return "", nil
+}
+
+// IsFlowSecure ensures the flow is secure and then, returns a JWT token.
+func (osvc *OIDCService) IsFlowSecure(ctx context.Context,
+	w http.ResponseWriter, r *http.Request) []byte {
+
+	token, err := osvc.claimsToken(ctx, w, r)
+	if err != nil {
+		return nil
+	}
+
+	session := osvc.sessionStorage.GetSession(r)
+	if session == nil {
+		msg := "An error occurred when trying to receive the session <nil>"
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusBadRequest)
+		return nil
+	}
+
+	nonceToken, err := osvc.GetValueFromToken("nonce", token)
+	if err != nil {
+		msg := fmt.Sprintf("An unexpected error occurred when getting the nonce from ID Token. Err: %v", err)
+		osvc.logger.Errorf(msg)
+		http.Error(w, msg, http.StatusBadRequest)
+		return nil
+	}
+	if !session.MatchNonce(nonceToken.(string), r) {
+		msg := "nonce did not match"
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusBadRequest)
+		return nil
+	}
+
+	jwt, err := osvc.createJWT(&token.CustomClaims)
+	if err != nil {
+		msg := fmt.Sprintf("An error occurred while creating a JWT. Error: %v", err)
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return nil
+	}
+
+	// If all goes well, then create the JSON data to display as body content.
+	jsonBody, err := json.Marshal(jwt)
+	if err != nil {
+		msg := fmt.Sprintf("An error occurred while validating some tokens. Error: %v", err)
+		osvc.logger.Errorf(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return nil
+	}
+
+	return jsonBody
+}
+
+// claimsToken creates token claims.
+func (osvc *OIDCService) claimsToken(ctx context.Context,
+	w http.ResponseWriter, r *http.Request) (*models.Tokens, error) {
```

Comment:
Pass the token as a parameter instead of `w http.ResponseWriter`, `r *http.Request`.